### PR TITLE
Implement passkey authentication for the back end

### PR DESCRIPTION
**Pull request by fritzmg**
This implements the ability to use passkeys / FIDO2 for the back end.

Within the 🛡️ Security section (user menu) there will be a new **Passkeys** section where you can click _Add passkey_

\
This will initialize the "attestation" process on your device (differing depending on browser and OS) which will create a passkey for the domain you are currently logged into the back end. After the passkey has been successfully created and submitted to the application, you will be prompted to enter a name for the passkey. If you do not give a name it will just show the internal ID of the passkey.

\
You can of course also create several passkeys, one for each device for example. Or delete any previously generated passkeys.

\
Once you have created one or more passkeys for your account, you can then use these passkeys to log into the backend - without having to specify a username. The passkey will uniquely identify your account.

\
There is also some basic error handling, e.g. if the attestation process fails or passkeys are not supported on your device.

### Notes

* Passkeys only work when you are connected via `https`.
* Chrome does not allow to create passkeys on domains with a self signed certificate.

### CODE REVIEW

File: core-bundle/contao/themes/flexible/styles/main.css [hunk 1]
```diff
@@ -1098,6 +1098,18 @@ h2.sub_headline {
 	line-height: 16px;
 }
 
+/* Passkeys */
+.passkeys .credential-list td.actions {
+	text-align: right;
+}
+
+.passkeys .credential-list td.actions button {
+	outline: 0;
```
Comment (leofeyer) [inline core-bundle/contao/themes/flexible/styles/main.css]: Is this okay with A11Y?

File: core-bundle/src/Entity/WebauthnCredential.php [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+#[Table(name: 'webauthn_credentials')]
+#[Entity(repositoryClass: WebauthnCredentialRepository::class)]
+class WebauthnCredential extends PublicKeyCredentialSource
+{
+    #[Column(type: Types::DATETIME_IMMUTABLE)]
+    public readonly \DateTimeImmutable $createdAt;
+
+    #[Id]
+    #[Column(unique: true)]
+    #[GeneratedValue(strategy: 'NONE')]
```
Comment (leofeyer) [inline core-bundle/src/Entity/WebauthnCredential.php:42]: Is this correct? Usually we have `'AUTO'`.

File: core-bundle/contao/templates/backend/be_login.html5 [hunk 3]
```diff
@@ -30,7 +30,7 @@
   <?php $this->block('container'); ?>
     <div id="container">
       <main id="main">
-        <form class="tl_login_form" method="post">
+        <form class="tl_login_form" method="post" data-controller="contao--webauthn" data-contao--webauthn-request-result-url-value="<?= $this->route('webauthn.controller.security.contao_backend.request.result') ?>" data-contao--webauthn-request-options-url-value="<?= $this->route('webauthn.controller.security.contao_backend.request.options') ?>" data-contao--webauthn-request-success-redirect-uri-value="<?= $this->webauthnSuccessUrl ?>">
```
Comment (leofeyer) [inline core-bundle/contao/templates/backend/be_login.html5]: As discussed, we want to add the button below the existing login form. Then the username and password fields can remain required.

File: core-bundle/contao/templates/backend/be_two_factor.html5 [hunk 4]
```diff
@@ -114,5 +114,43 @@
       <a href="<?= $this->route('contao_backend', ['do' => 'security', 'act' => 'enable', 'ref' => $this->ref]) ?>" class="tl_submit"><?= $this->trans('MSC.enable') ?></a>
     </div>
   <?php endif; ?>
+</div>
 
+<div class="two-factor passkeys">
+  <h2 class="sub_headline">Passkeys</h2>
+  <p>Here you can add and manage your passkeys with which you can log into the back end. Passkeys allow you to sign in safely and easily, without requiring a password and two-factor authentication.</p>
```
Comment (leofeyer) [inline core-bundle/contao/templates/backend/be_two_factor.html5]: Just a reminder to make all this translatable.

File: manager-bundle/skeleton/config/security.yaml [hunk 5]
```diff
@@ -17,6 +17,13 @@ security:
             user_checker: contao.security.user_checker
             switch_user: true
             login_throttling: ~
+            webauthn:
```
Comment (leofeyer) [inline manager-bundle/skeleton/config/security.yaml:20]: We have to add this to the `README.md` and `core-bundle/tests/Functional/app/config/security.yaml` files, too.

File: core-bundle/src/Entity/WebauthnCredential.php (hunk 2 above)
Comment (fritzmg) [reply core-bundle/src/Entity/WebauthnCredential.php:42]: This was taken straight from the [example](https://webauthn-doc.spomky-labs.com/symfony-bundle/entities-with-doctrine#the-doctrine-entity). This way you can have uuids as the primary key (generated in the constructor below).

**Comment by leofeyer** [comment]
This is freaking awesome 🚀

File: core-bundle/contao/languages/en/default.xlf [hunk 6]
```diff
@@ -2189,6 +2189,9 @@
       <trans-unit id="MSC.altchaWaitAlert">
         <source>Verifying … please wait.</source>
       </trans-unit>
+      <trans-unit id="MSC.passkeyLogin">
+        <source>Login with Passkey</source>
```
Comment (leofeyer) [inline core-bundle/contao/languages/en/default.xlf]:
```suggestion
        <source>Sign in with a passkey</source>
```

Like GitHub does?

**Comment by leofeyer** [comment]
I have noticed two minor issues:

### URL does not change

* Log into the back end, go to "user" → "security" and add a passkey.
* After naming the passkey, the URL still has the `&edit_new_passkey=1` fragment, so reloading the page will open the naming dialog again.

### Focus is lost?

* Log into the back end, go to "user" → "security" and edit a passkey.
* After saving the new name, the user menu in the orange top bar can no longer be opened.

**Comment by fritzmg** [comment]
> After naming the passkey, the URL still has the `&edit_new_passkey=1` fragment, so reloading the page will open the naming dialog again.

Ah yes, I meant to change that, it's an easy fix with an attribute on the Turbo Frame. 👍 

> After saving the new name, the user menu in the orange top bar can no longer be opened.

Hm, I was not able to reproduce that. But it could be some kind of race condition that would also happen with any other Turbo requests though, so likely not related to the passkey editing directly.

Regardless, we should probably implement the JavaScript for the user menu as either a Stimulus Controller or a Mutation Observer. Do you want me to add this to this PR or rather create a separate PR for that?

**Comment by fritzmg** [comment]
> > After naming the passkey, the URL still has the `&edit_new_passkey=1` fragment, so reloading the page will open the naming dialog again.
> 
> Ah yes, I meant to change that, it's an easy fix with an attribute on the Turbo Frame. 👍

This is now fixed.

**Comment by fritzmg** [comment]
The PHP 8.1 CI fails because with this PR we now need at least PHP 8.2. Should I adjust this in this PR as well?

**Comment by leofeyer** [comment]
Yes please.

> Regardless, we should probably implement the JavaScript for the user menu as either a Stimulus Controller or a Mutation Observer. Do you want me to add this to this PR or rather create a separate PR for that?

And also add this please 😊

**Comment by fritzmg** [comment]
Both done 👍

**Comment by fritzmg** [comment]
Regarding code coverage: we currently do not write tests for repositories. Should I add tests here?

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @fritzmg.

**Pull request merged**